did/ion/request.go

```diff
@@ -0,0 +1,302 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign JWT: %w", err)
+	}
+	return &DeactivateRequest{
+		Type:        Deactivate,
+		DIDSuffix:   didSuffix,
+		RevealValue: revealValue,
+		SignedData:  signedJWT,
+	}, nil
+}
+
+// NewRecoverRequest creates a new recover request https://identity.foundation/sidetree/spec/#recover
+func NewRecoverRequest(didSuffix string, recoveryKey, nextRecoveryKey, nextUpdateKey crypto.PublicKeyJWK, document Document, signer BTCSignerVerifier) (*RecoverRequest, error) { //revive:disable-line:argument-limit
+	// prepare reveal value
+	revealValue, _, err := Commit(recoveryKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// prepare delta
+	patches := []any{
+		ReplaceAction{
+			Action:   Replace,
+			Document: document,
+		},
+	}
+
+	_, updateCommitment, err := Commit(nextUpdateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	delta := Delta{
+		UpdateCommitment: updateCommitment,
+		Patches:          patches,
+	}
+
+	// prepare signed data
+	deltaCanonical, err := CanonicalizeAny(delta)
+	if err != nil {
+		return nil, err
+	}
+	deltaHash, err := HashEncode(deltaCanonical)
+	if err != nil {
+		return nil, err
+	}
+	_, recoveryCommitment, err := Commit(nextRecoveryKey)
+	if err != nil {
+		return nil, err
+	}
+
+	toBeSigned := struct {
+		RecoveryCommitment string              `json:"recoveryCommitment"`
+		RecoveryKey        crypto.PublicKeyJWK `json:"recoveryKey"`
+		DeltaHash          string              `json:"deltaHash"`
+	}{
+		RecoveryCommitment: recoveryCommitment,
+		RecoveryKey:        recoveryKey,
+		DeltaHash:          deltaHash,
+	}
+	signedJWT, err := signer.SignJWT(toBeSigned)
+	if err != nil {
+		return nil, err
+	}
+	return &RecoverRequest{
+		Type:        Recover,
+		DIDSuffix:   didSuffix,
+		RevealValue: revealValue,
+		Delta:       delta,
+		SignedData:  signedJWT,
+	}, nil
+}
+
+type StateChange struct {
+	ServicesToAdd        []Service
+	ServiceIDsToRemove   []string
+	PublicKeysToAdd      []PublicKey
+	PublicKeyIDsToRemove []string
+}
+
+func (s StateChange) IsValid() error {
+	// check if services are valid
+	// build index of services to make sure IDs are unique
+	services := make(map[string]Service)
+	for _, service := range s.ServicesToAdd {
+		if _, ok := services[service.ID]; ok {
+			return fmt.Errorf("service %s duplicated", service.ID)
+		}
+
+		if len(service.ID) > maxIDLength {
+			return fmt.Errorf("service<%s> id is too long", service.ID)
+		}
+
+		// make sure service is valid if it's not a dupe
+		if len(service.Type) > maxServiceTypeLength {
+			return fmt.Errorf("service<%s> type %s is too long", service.ID, service.Type)
+		}
+
+		services[service.ID] = service
+	}
+
+	// check if public keys are valid
+	// build index of public keys to add
+	publicKeys := make(map[string]PublicKey)
+	for _, publicKey := range s.PublicKeysToAdd {
+		if _, ok := publicKeys[publicKey.ID]; ok {
+			return fmt.Errorf("public key<%s> is duplicated", publicKey.ID)
```

Comment:
Here and elsewhere. If you want stack traces, use `errors.Errorf`. See https://stackoverflow.com/questions/61933650/whats-the-difference-between-errors-wrapf-errors-errorf-and-fmt-errorf for more details. (make sure it's **not** the native `"errors"` library).